lib/serverengine/socket_manager.rb

```diff
@@ -0,0 +1,125 @@
+#
+# ServerEngine
+#
+# Copyright (C) 2012-2013 Sadayuki Furuhashi
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'socket'
+require 'fcntl'
+
+module ServerEngine
+  module SocketManager
+
+    def self.new_socket_manager
+      if $platformwin
+        # TODO: using ARGV temporarily. I plan to create interface class of command parser and use it.
+        drb_uri = ARGV[-1]
+        begin
+          SocketManagerWin::Client.new(drb_uri)
+        rescue
+          ServerEngine.dump_uncaught_error($!)
+        end
+      else
+        # TODO: using ARGV temporarily. I plan to create interface class of command parser and use it.
+        uds_drb = ARGV[-1]
+        unix_socket_client = UNIXSocket.for_fd(uds_drb.split('#').first.to_i)
+        drb_uri = uds_drb.split('#').last
+        begin
+          SocketManager::Client.new(unix_socket_client, drb_uri)
+        rescue
+          ServerEngine.dump_uncaught_error($!)
+        end
+      end
+    end
+
+    class Client
+      def initialize(unix_socket_client, drb_uri)
+        @unix_socket_client = unix_socket_client
+        @sm_server = DRb::DRbObject.new_with_uri(drb_uri)
+      end
+
+      def get_udp(bind, port)
+        @sm_server.udp_socket_fd(bind, port)
+        @unix_socket_client.recv_io
+      end
+
+      def get_tcp(bind, port)
+        @sm_server.socket_fd(bind, port)
+        @unix_socket_client.recv_io
+      end
+    end
+
+    class Server
+      def initialize
+        @tcp_socks = {}
+        @udp_socks = {}
+        @unix_socket_server = nil
+      end
+
+      def new_unix_socket
+        @unix_socket_server, unix_socket_client = UNIXSocket.pair
+        unix_socket_client
+      end
+
+      def socket_fd(bind, port)
+
+        socks_key = bind.to_s + port.to_s
+
+        if @tcp_socks.has_key?(socks_key)
```

Comment:
For now, sock_fd will not be called by multiple threads, but if it will, I think DRb can deal with this.
If I should take measures, I'll try to implement with BlockingFlag.